safe_control_gym/experiment.py

```diff
@@ -1,148 +1,238 @@
+            "average_return": np.asarray(met.get_episode_returns()).mean(),
+            "average_rmse": np.asarray(met.get_episode_rmse()).mean(),
+            "rmse_std": np.asarray(met.get_episode_rmse()).std(),
+            "worst_case_rmse_at_0.5": compute_cvar(np.asarray(met.get_episode_rmse()), 0.5, lower_range=False),
+            "failure_rate":  np.asarray(met.get_episode_constraint_violations()).mean(),       
+            "average_constraint_violation": np.asarray(met.get_episode_constraint_violation_steps()).mean(),
             # others ???
         }
-        return results
+        return metrics
     
+    def reset(self):
+        """Resets the environments, controller, and safety filter to prepare for training or evaluation. """
+
+        self.env.reset()
+        self.ctrl.reset()
+
+        if self.safety_filter is not None:
+            self.safety_filter.reset()
+    
+    def close(self):
+        """Closes the environments, controller, and safety filter. """
+
+        self.env.close()
+        self.ctrl.close()
+
+        if self.safety_filter is not None:
+            self.safety_filter.close()
+    
+    def load(self, ctrl_path=None, safety_filter_path=None):
+        """Restores model of the controller and/or safety filter given checkpoint paths. 
+        
+        Args:
+            ctrl_path (str): the path used to load the controller's model
+            safety_filter_path (str): the path used to load the safety_filter's model
+        """
+
+        if ctrl_path is not None:
+            self.ctrl.load(ctrl_path)
+        if safety_filter_path is not None:
+            self.safety_filter.load(safety_filter_path)
+    
+    def save(self, ctrl_path=None, safety_filter_path=None):
+        """Saves the model of the controller and/or safety filter given checkpoint paths. 
+        
+        Args:
+            ctrl_path (str): the path used to save the controller's model
+            safety_filter_path (str): the path used to save the safety_filter's model
+        """
+
+        if ctrl_path is not None:
+            self.ctrl.save(ctrl_path)
+        if safety_filter_path is not None:
+            self.safety_filter.save(safety_filter_path)
     
 
 class RecordDataWrapper(gym.Wrapper):
     """A wrapper to standardizes logging for benchmark envs.
     
     currently saved info
-    * obs, rew, done, info, act
-    * env.state, env.current_preprocessed_action
-    
+    * obs, reward, done, info, action
+    * env.state, env.current_raw_input_action, env.current_preprocessed_action
     """
+
     def __init__(self, env, deque_size=None, **kwargs):
         super().__init__(env)
-        self.deque_size = deque_size
-        self.initialize_data_containers()
-
-    def initialize_data_containers(self):
-        """"""
         self.episode_data = defaultdict(list)
-        self.data = defaultdict(lambda: deque(self.deque_size))
+        self.data = defaultdict(list)
```

Comment:
food for thought here: the reason for using `deque` is to consider logging only the latest `deque_size` number of episode data during training, otherwise we might be caching old data as well (unless there's also reset of `env.data` after each metric evaluation call).

Reply:
Hmm maybe its cleaner to have a reset function that resets env.data since I think its kind of weird that the data would just randomly get full. In experiment.reset() we could clear the env.data

Reply:
if we don't want it to get full, can set `deque_size` to None so it's the same as a list. 

two following questions are 
- is `experiment.reset()` meant to be called once per training or many times before each `run_evaluation()` call?  (related to the general question #2 at the bottom) 
- are we trying to log data from the training env as well? and if so how?    (this is coz the wrapper and `deque` were originally designed to log training env data, say ppo wants to check the latest 10 training episodes' data or the current total number of constraint violations so far)

Reply:
I added a flag in run_evaluation that does not reset the controller if it is training. Also added a clear_data function to the environment that can be called when reseting the entire experiment. I have left it as a list since idk the benefit of a deque if it has no size and we are only appending to the end, but if you think its better we can switch to deque. I don't think this solves exactly how we will do RL training, but I think it addresses the immediate comments

Reply:
For RL, need top 10 items or something like that. Will change for RL